polaris-service/src/main/java/io/polaris/service/catalog/BasePolarisCatalog.java

```diff
@@ -846,10 +847,51 @@ private void validateLocationForTableLike(
       TableIdentifier identifier,
       String location,
       PolarisResolvedPathWrapper resolvedStorageEntity) {
-    Optional<PolarisStorageConfigurationInfo> optStorageConfiguration =
-        PolarisStorageConfigurationInfo.forEntityPath(
-            callContext.getPolarisCallContext().getDiagServices(),
-            resolvedStorageEntity.getRawFullPath());
+
+    boolean allowTableLocationOutsideNamespaceLocation =
+        callContext
+            .getPolarisCallContext()
+            .getConfigurationStore()
+            .getConfiguration(
+                callContext.getPolarisCallContext(),
+                PolarisConfiguration.ALLOW_TABLE_LOCATION_OUTSIDE_NAMESPACE_LOCATION);
+
+    boolean allowTableLocationOutsideCatalogLocation =
+        callContext
+            .getPolarisCallContext()
+            .getConfigurationStore()
+            .getConfiguration(
+                callContext.getPolarisCallContext(),
+                PolarisConfiguration.ALLOW_TABLE_LOCATION_OUTSIDE_CATALOG_LOCATION);
+
+    if (allowTableLocationOutsideCatalogLocation && !allowTableLocationOutsideNamespaceLocation) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Incompatible configuration for %s: `%s` is set but `%s` is not",
+              identifier.name(),
+              PolarisConfiguration.ALLOW_TABLE_LOCATION_OUTSIDE_CATALOG_LOCATION.key,
+              PolarisConfiguration.ALLOW_TABLE_LOCATION_OUTSIDE_NAMESPACE_LOCATION.key));
+    }
+
+    Optional<PolarisStorageConfigurationInfo> optStorageConfiguration = Optional.empty();
+    if (allowTableLocationOutsideNamespaceLocation) {
+      if (allowTableLocationOutsideCatalogLocation) {
+        optStorageConfiguration = Optional.empty();
+      } else {
+        optStorageConfiguration =
+            findStorageInfoFromHierarchy(resolvedStorageEntity)
+                .map(
+                    storageInfoHolderEntity -> {
+                      return new CatalogEntity(storageInfoHolderEntity)
+                          .getStorageConfigurationInfo();
+                    });
+      }
+    } else {
+      optStorageConfiguration =
+          PolarisStorageConfigurationInfo.forEntityPath(
+              callContext.getPolarisCallContext().getDiagServices(),
+              resolvedStorageEntity.getRawFullPath());
+    }
```

Comment:
Why is this logic here rather than in `PolarisStorageConfigurationInfo`? That's where the other storage configuration override code is